transport/tchannel/header.go

```diff
@@ -236,10 +251,46 @@ func getHeaderMap(hs transport.Headers, headerCase headerCase) map[string]string
 	}
 }
 
-func deleteReservedHeaders(headers transport.Headers) {
+func findReservedHeaderPrefix(headers map[string]string) (string, bool) {
+	for key := range headers {
+		if isReservedHeaderPrefix(key) {
+			return key, true
+		}
+	}
+	return "", false
+}
+
+func validateApplicationHeaders(headers map[string]string, edgeMetrics observability.ReservedHeaderEdgeMetrics) error {
+	key, found := findReservedHeaderPrefix(headers)
+	if !found {
+		return nil
+	}
+
+	edgeMetrics.IncError()
+
+	if enforceHeaderRules {
+		return yarpcerrors.InternalErrorf("header with rpc prefix is not allowed in request application headers (%s was passed)", key)
+	}
+	return nil
+}
+
+func deleteReservedHeaders(headers transport.Headers, edgeMetrics observability.ReservedHeaderEdgeMetrics) {
 	for headerKey := range _reservedHeaderKeys {
 		headers.Del(headerKey)
 	}
+
+	deleteReservedPrefixHeaders(headers, edgeMetrics)
+}
+
+func deleteReservedPrefixHeaders(headers transport.Headers, edgeMetrics observability.ReservedHeaderEdgeMetrics) {
+	for key := range headers.Items() {
+		if isReservedHeaderPrefix(key) {
```

Comment:
let's reduce some nesting here:
```go
if !isReservedHeaderPrefix(key) {
   continue
}
edgeMetrics.IncStripped()
if enforceHeaderRules {
	headers.Del(key)
}
```

another question, I don't see reuse here. Can we just combine those 2 and make a single `deleteReservedHeaders` func?

Reply:
Fixed

> another question, I don't see reuse here
It's reused here: https://github.com/yarpc/yarpc-go/pull/2272/files#diff-72c944f1bd46d354d3c1255093274ae1ec4fca8fef63abdcd2fbc0f6ee738e86R188